### ISSUE TITLE
Memoizing queries for faster performance

### ISSUE THREAD

**Issue by benoror** [ecaresoft/taxi-driver#19]
Cache expensive match queries using [Memoization](https://en.wikipedia.org/wiki/Memoization) technique

> Re-using a lambda means the underlying container that hosts it does not get discarded. State gets preserved (such as temporary files, memory caches, sub-processes). This empowers the developer not just to minimize the time spent in the booting process, but to also take advantage of caching data (in memory or /tmp filesystem) and memoizing expensive computations.

https://zeit.co/docs/v2/deployments/concepts/lambdas

### Resources / eamples:

- https://github.com/zeit/hyper/blob/a5177e8df847dcb1c0efa20f97f3858613972ef1/cli/api.js#L21-L31
- https://medium.freecodecamp.org/understanding-memoize-in-javascript-51d07d19430e
- https://addyosmani.com/blog/faster-javascript-memoization/
- https://codeburst.io/understanding-memoization-in-3-minutes-2e58daf33a19
- https://www.npmjs.com/package/fast-memoize
  - https://community.risingstack.com/the-worlds-fastest-javascript-memoization-library/
- https://caolan.github.io/async/memoize.js.html

**Comment by pabloemilio97** [comment]
Did a commit on optimizeBestMatch PR for this issue https://github.com/ecaresoft/taxi-driver/pull/22/commits/37aec12252d7039637e083baa2579406ee7c75ee

**Comment by benoror** [comment]
> Problem: yarn tests only succeed if `cache.flush()` is executed each time `getTaxes()` is called (don't know why!) Nevertheless, cache hits happen when the same call is made twice, and results are retrieved faster. @benoror thoughts?

@pabloemilio97 Good catch! AFAIK flushing any kind of cache is a good practice for testing, to ensure every test is "deterministic".

I did something similar using the `beforeEach` hook for these tests:

- https://github.com/ecaresoft/taxi-driver/blob/9fec4c080da4be50113bf273178a0f7f43782795/lib/eval.test.js#L5-L7
- https://github.com/ecaresoft/taxi-driver/blob/f969d39f90ca05f7e15ba3c3439c7e1e6f67cc75/lib/calc.test.js#L86-L88
- https://github.com/ecaresoft/taxi-driver/blob/f969d39f90ca05f7e15ba3c3439c7e1e6f67cc75/lib/calc.test.js#L52-L54

In this context the Formula Parser would "cache" variables (https://github.com/handsontable/formula-parser), but the same principle applies to any caching scheme, such as you `node-cache` example.

Maybe you could ask @luissifu to take a look at this and have his opinion on the matter?

**Comment by benoror** [comment]
> make it store cache for each of the functions

Will `node-cache` store the cache in-memory? or in disk? Zeit lambdas have some specific requierements for in-disk files (tl;dr; use `/tmp`): https://zeit.co/docs/v2/deployments/concepts/lambdas#filesystem
